File: web/src/api/events.ts
// Wire-format event types mirroring backend src/agentbox/api/events.py.
// Single source of truth for the WS stream + transcript shape.

export const EVENT_TYPES = [
  'text',
  'log',
  'tool_call',
  'tool_result',
  'usage',
  'retry',
  'thinking',
  'timeout',
  'validation',
  'done',
] as const;

export type EventType = (typeof EVENT_TYPES)[number];

interface EventBase {
  ts?: string;
  run_id?: string;
}

export interface TextEvent extends EventBase {
  type: 'text';
  text: string;
  role?: 'assistant' | 'user' | 'system';
  delta?: boolean;
}

export interface LogEvent extends EventBase {
  type: 'log';
  level?: 'debug' | 'info' | 'warn' | 'error';
  message: string;
}

export interface ToolCallEvent extends EventBase {
  type: 'tool_call';
  tool: string;
  arguments: Record<string, unknown>;
  call_id?: string | null;
}

export interface ToolResultEvent extends EventBase {
  type: 'tool_result';
  tool: string;
  call_id?: string | null;
  ok?: boolean;
  result_excerpt?: string | null;
}

export interface UsageEvent extends EventBase {
  type: 'usage';
  input_tokens?: number;
  output_tokens?: number;
  cache_read_tokens?: number;
  cache_write_tokens?: number;
  cost_usd?: number | null;
  model?: string | null;
}

export interface RetryEvent extends EventBase {
  type: 'retry';
  attempt?: number;
  reason?: string;
  error?: string | null;
}

export interface ThinkingEvent extends EventBase {
  type: 'thinking';
  text: string;
}

export interface TimeoutEvent extends EventBase {
  type: 'timeout';
  timeout_seconds?: number;
  error?: string | null;
}

export interface ValidationEvent extends EventBase {
  type: 'validation';
  ok: boolean;
  attempt?: number;
  mode?: 'strict' | 'warn' | 'off';
  engine?: 'jsonschema' | 'pydantic' | 'both' | 'none';
  error?: string | null;
}

export interface DoneEvent extends EventBase {
  type: 'done';
  ok: boolean;
  exit_code?: number | null;
  error?: string | null;
  status?: 'ok' | 'error' | 'timeout' | null;
}

export type RunEvent =
  | TextEvent
  | LogEvent
  | ToolCallEvent
  | ToolResultEvent
  | UsageEvent
  | RetryEvent
  | ThinkingEvent
  | TimeoutEvent
  | ValidationEvent
  | DoneEvent;

// UI extension flags added in-memory by the event-stream coalescer.
// Not part of the wire format — kept here so consumers don't redeclare them.
export interface CoalescedFlags {
  _coalesced?: boolean;
  _coalesced_count?: number;
}

export type UiRunEvent = RunEvent & CoalescedFlags;

// Loose shape for UI code that indexes arbitrary fields without
// narrowing on `type`. Prefer `RunEvent` when you can.
export interface LooseStreamEvent extends CoalescedFlags {
  type: string;
  [k: string]: unknown;
}

// Visual metadata for each event type. Centralized so EventStream,
// ConversationView, and any future viewer share the same palette.
export interface EventMeta {
  label: string;
  color: string;
  bg: string;
}

export const EVENT_META: Record<EventType, EventMeta> = {
  text:        { label: 'text',       color: '#c9d1d9', bg: 'rgba(201,209,217,0.08)' },
  log:         { label: 'log',        color: '#8b949e', bg: 'rgba(139,148,158,0.08)' },
  tool_call:   { label: 'call',       color: '#79c0ff', bg: 'rgba(121,192,255,0.10)' },
  tool_result: { label: 'result',     color: '#3fb950', bg: 'rgba(63,185,80,0.10)' },
  usage:       { label: 'usage',      color: '#d29922', bg: 'rgba(210,153,34,0.10)' },
  retry:       { label: 'retry',      color: '#f0883e', bg: 'rgba(240,136,62,0.10)' },
  thinking:    { label: 'thinking',   color: '#58a6ff', bg: 'rgba(88,166,255,0.10)' },
  timeout:     { label: 'timeout',    color: '#f0883e', bg: 'rgba(240,136,62,0.15)' },
  validation:  { label: 'validation', color: '#d2a8ff', bg: 'rgba(210,168,255,0.10)' },
  done:        { label: 'done',       color: '#58a6ff', bg: 'rgba(88,166,255,0.10)' },
};

const UNKNOWN_META: EventMeta = {
  label: 'unknown',
  color: 'var(--fg-muted)',
  bg: 'rgba(139,148,158,0.06)',
};

export function getEventMeta(type: string): EventMeta {
  return (EVENT_META as Record<string, EventMeta>)[type] ?? { ...UNKNOWN_META, label: type };
}

// One-line summary of an event, used in compact list rows.
export function summarizeEvent(ev: LooseStreamEvent): string {
  switch (ev.type) {
    case 'text':
      return String(ev.text ?? '').slice(0, 180);
    case 'log':
      return `[${ev.level ?? ''}] ${String(ev.message ?? '').slice(0, 180)}`;
    case 'tool_call':
      return `${ev.tool ?? ''} ${JSON.stringify(ev.arguments ?? {}).slice(0, 140)}`;
    case 'tool_result':
      return `${ev.tool ?? ''} ok=${ev.ok}`;
    case 'usage':
      return `in=${ev.input_tokens ?? 0} out=${ev.output_tokens ?? 0} cost=$${ev.cost_usd ?? 0}`;
    case 'done':
      return `status=${ev.status ?? ''}`;
    case 'retry':
      return `attempt=${ev.attempt ?? ''} reason=${ev.reason ?? ''}`;
    case 'thinking':
      return String(ev.text ?? '').slice(0, 180);
    case 'timeout':
      return `timeout after ${ev.timeout_seconds ?? '?'}s`;
    case 'validation':
      return `ok=${ev.ok} engine=${ev.engine ?? ''} mode=${ev.mode ?? ''} attempt=${ev.attempt ?? ''}${ev.error ? ` — ${String(ev.error).slice(0, 120)}` : ''}`;
    default:
      return JSON.stringify(ev).slice(0, 180);
  }
}

File: src/agentbox/core/data/runs/events.py
"""RunEvent discriminated union — single source of truth for the wire protocol.

Every event is a Pydantic model with a discriminator field ``type``. The
union ``RunEvent`` is the wire format for the WS stream and the on-disk
transcript (one JSON object per line).

**This is the canonical home.** ``core/data/events.py`` and ``api/events.py``
are thin re-exports for backward compatibility. New code MUST import from
here (or from the ``core.execution`` facade once it lands).
"""

from __future__ import annotations

from datetime import UTC, datetime
from typing import Annotated, Literal

from pydantic import BaseModel, Field

from agentbox.core.constants import EventType


class _EventBase(BaseModel):
    ts: datetime = Field(default_factory=lambda: datetime.now(UTC))
    run_id: str


class ToolCallEvent(_EventBase):
    type: Literal[EventType.TOOL_CALL] = EventType.TOOL_CALL
    tool: str
    arguments: dict
    call_id: str | None = None


class ToolResultEvent(_EventBase):
    type: Literal[EventType.TOOL_RESULT] = EventType.TOOL_RESULT
    tool: str
    call_id: str | None = None
    ok: bool = True
    result_excerpt: str | None = None
    """First N chars of the result. Full payload lives in run artifacts."""


class TextEvent(_EventBase):
    type: Literal[EventType.TEXT] = EventType.TEXT
    text: str
    role: Literal["assistant", "user", "system"] = "assistant"
    # True when this event is an incremental chunk emitted live during
    # the run. The executor skips appending deltas to ``output_text`` —
    # a non-delta TextEvent with the consolidated/cleaned text is
    # expected after the runner finishes.
    delta: bool = False


class UsageEvent(_EventBase):
    type: Literal[EventType.USAGE] = EventType.USAGE
    input_tokens: int = 0
    output_tokens: int = 0
    cache_read_tokens: int = 0
    cache_write_tokens: int = 0
    cost_usd: float | None = None
    model: str | None = None


class LogEvent(_EventBase):
    type: Literal[EventType.LOG] = EventType.LOG
    level: Literal["debug", "info", "warn", "error"] = "info"
    message: str


class RetryEvent(_EventBase):
    type: Literal[EventType.RETRY] = EventType.RETRY
    attempt: int = 1
    reason: str = ""
    """Why the retry happened: 'validation_failed' | 'run_error' | 'timeout'."""
    error: str | None = None
    """The error or validation message that triggered the retry."""


class ThinkingEvent(_EventBase):
    type: Literal[EventType.THINKING] = EventType.THINKING
    text: str
    """Raw thinking / reasoning content from the model."""


class TimeoutEvent(_EventBase):
    type: Literal[EventType.TIMEOUT] = EventType.TIMEOUT
    timeout_seconds: int = 0
    """The configured timeout for this run."""
    error: str | None = None


class ValidationEvent(_EventBase):
    type: Literal[EventType.VALIDATION] = EventType.VALIDATION
    ok: bool
    attempt: int = 1
    mode: Literal["strict", "warn", "off"] = "strict"
    engine: Literal["jsonschema", "pydantic", "both", "none"] = "none"
    error: str | None = None
    """Validation error message when ok=False."""


class DoneEvent(_EventBase):
    type: Literal[EventType.DONE] = EventType.DONE
    ok: bool
    exit_code: int | None = None
    error: str | None = None
    # Specific failure category. ``"timeout"`` means the runner's
    # ``timeout_seconds`` expired before the process exited. Any other
    # non-ok run is recorded as ``"error"``. None when ``ok=True``.
    status: Literal["ok", "error", "timeout"] | None = None


RunEvent = Annotated[
    ToolCallEvent
    | ToolResultEvent
    | TextEvent
    | UsageEvent
    | LogEvent
    | RetryEvent
    | ThinkingEvent
    | TimeoutEvent
    | ValidationEvent
    | DoneEvent,
    Field(discriminator="type"),
]

__all__ = [
    "DoneEvent",
    "LogEvent",
    "RetryEvent",
    "RunEvent",
    "TextEvent",
    "ThinkingEvent",
    "TimeoutEvent",
    "ToolCallEvent",
    "ToolResultEvent",
    "UsageEvent",
    "ValidationEvent",
]
